### PR TITLE
Use unchecked for ERC721 balance updates

### PR DESCRIPTION
**Pull request by Amxx**
Following the logic in described in https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3512#issuecomment-1172614573.

- `_balances` are private
- `_balances` are only increased/decreased when tokens are minted, burned, or transferred

A balance cannot decrease below 0 because that would require transferring/burning more tokens out of a wallet than were transferred/minted to this wallet in the first place.

The balance cannot overflow, because that would require minting all the tokenIds, and putting them all in the same wallet (there are 2**256 tokenIds, so that would overflow by one). While this is mathematically possible, it is impossible in practice (with token being transferred/minted one at a time)

#### PR Checklist

- [ ] Tests
- [ ] Documentation
- [ ] Changelog entry

### CODE REVIEW

**Comment by frangio** [comment]
Given that we want to explore batch minting in ERC721 it seems we should not merge this, as the assumption that minting happens one at a time will be invalidated. Also please include benchmarks, what kind of improvement do you see with this PR?

**Comment by Amxx** [comment]
I understand that we want to experiment new minting mechanism. However, I believe that these minting mechanisms should remain ERC721 compliant, which means emitting a Transfer event for all the tokens ids. This means that, in practice, it's impossible to mint all the tokens, and the balance increment overflow is not an issue.

**Comment by Amxx** [comment]
For reference: https://twitter.com/optimizoor/status/1544626818996613120?t=8tV84gD3blTSsxk14EotiQ

TLDR: batch minting should be limited to "not to big batches" ... which means that the limit should never be reached, even minting 10**18 token each time

File: contracts/token/ERC721/ERC721.sol
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol:359]: `_beforeTokenTransfer` could in theory move the token away and this would result in overflow if we use unchecked.

```suggestion
        _balances[from] -= 1;
        unchecked {
            // `_balances[to]` could overflow in the conditions described in `_mint`. That would require
            // all 2**256 token ids to be minted, which in practice is impossible.
```

File: contracts/token/ERC721/ERC721.sol
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol:321]: For the same reason, because of `_beforeTokenTransfer` we can't use unchecked here.

```suggestion
        _balances[owner] -= 1;
```

**Review by frangio — CHANGES_REQUESTED** [review]

File: contracts/token/ERC721/ERC721.sol
Comment (Amxx) [reply contracts/token/ERC721/ERC721.sol:359]: If `_beforeTokenTransfer` transfers the token, then the balances would be invalid regardless of the `unchecked`. This is a wider issue, that we should address independently, and that IMO should not prevent the use of unchecked here.

**Comment by Amxx** [comment]
I believe once #3611 is merged, the requested changes will no longer be needed.

**Review by frangio — APPROVED** [review]
LGTM

**Pull request merged**
